### PR TITLE
Implicit memory aliasing in for loop.

### PR DESCRIPTION
**Pull request by slayer321**
Signed-off-by: slayer321 <sachin.maurya7666@gmail.com>
no issue related to bug bash
https://lift.sonatype.com/result/longhorn/longhorn-manager/01FHW2JKTWJ3Z4RA5MMT2A53WS?t=CustomTool%20%22GoSec%22%7CG601

### CODE REVIEW

File: controller/engine_image_controller.go [hunk 1]
```diff
@@ -667,9 +667,9 @@ func (ic *EngineImageController) enqueueControlleeChange(obj interface{}) {
 		return
 	}
 	ownerRefs := metaObj.GetOwnerReferences()
-	for _, ref := range ownerRefs {
+	for i, _ := range ownerRefs {
```
[the remark this answers is not in the document]
Comment (slayer321) [reply controller/engine_image_controller.go]: ignore

**Review by shuo-wu — COMMENTED** [review]
Thanks for your contribution! Can you squash the 2nd ci fix commit into the 1st commit? Others LGTM

File: app/upgrader.go [hunk 2]
```diff
@@ -22,13 +23,13 @@ func upgradeLonghornRelatedComponents(kubeClient *clientset.Clientset, namespace
 	if err != nil {
 		return err
 	}
-	for _, pv := range pvList.Items {
+	for i, pv := range pvList.Items {
 		// Make sure we are handling the Longhorn related PVs only
 		if v, exist := pv.Annotations[PVAnnotationCSIProvisioner]; !exist || v != types.DeprecatedProvisionerName {
 			continue
 		}
 		pv.Annotations[PVAnnotationCSIProvisioner] = types.LonghornDriverName
```
Comment (shuo-wu) [inline app/upgrader.go:31]: It seems that we need to modify this as well:
```
pvList.Items[i].Annotations[PVAnnotationCSIProvisioner] = types.LonghornDriverName
```

or using this kind of format:
```
for i := range pvList.Items {
	pv := &pvList.Items[i]
	if v, exist := pv.Annotations[PVAnnotationCSIProvisioner]; !exist || v != types.DeprecatedProvisionerName {
		continue
	}
	pv.Annotations[PVAnnotationCSIProvisioner] = types.LonghornDriverName

	if _, err := kubeClient.CoreV1().PersistentVolumes().Update(context.TODO(), pv, metav1.UpdateOptions{}); err != nil {
		return err
	}
```

The key point of these modifications is avoiding using the pointer of the temporary variable initialized by the "for" loop.

File: upgrade/v070to080/upgrade.go [hunk 3]
```diff
@@ -68,11 +68,11 @@ func doInstanceManagerUpgrade(namespace string, lhClient *lhclientset.Clientset)
 		}
 		return errors.Wrapf(err, upgradeLogPrefix+"failed to list all existing instance managers during the instance managers upgrade")
 	}
-	for _, im := range imList.Items {
+	for i, im := range imList.Items {
 		if im.Spec.Image != "" {
 			continue
 		}
-		im := &im
+		im := &imList.Items[i]
```
Comment (shuo-wu) [inline upgrade/v070to080/upgrade.go:75]:
```suggestion
	for i := range imList.Items {
		im := &imList.Items[i]
		if im.Spec.Image != "" {
			continue
		}
```

File: upgrade/v102to110/upgrade.go [hunk 4]
```diff
@@ -109,13 +109,13 @@ func upgradeVolumes(namespace string, lhClient *lhclientset.Clientset) (err erro
...
 
-	for _, v := range volumeList.Items {
+	for i, v := range volumeList.Items {
 		// in pr https://github.com/longhorn/longhorn-manager/pull/789
 		// we added a new access mode field, that is exposed to the ui
 		// so we add the previously only supported rwo access mode
 		if v.Spec.AccessMode == "" {
 			v.Spec.AccessMode = types.AccessModeReadWriteOnce
-			updatedVolume, err := lhClient.LonghornV1beta1().Volumes(namespace).Update(context.TODO(), &v, metav1.UpdateOptions{})
+			updatedVolume, err := lhClient.LonghornV1beta1().Volumes(namespace).Update(context.TODO(), &volumeList.Items[i], metav1.UpdateOptions{})
```
Comment (shuo-wu) [inline upgrade/v102to110/upgrade.go:118]: Similar here:
```suggestion
	for i := range volumeList.Items {
		v := &volumeList.Items[i]
		// in pr https://github.com/longhorn/longhorn-manager/pull/789
		// we added a new access mode field, that is exposed to the ui
		// so we add the previously only supported rwo access mode
		if v.Spec.AccessMode == "" {
			v.Spec.AccessMode = types.AccessModeReadWriteOnce
			updatedVolume, err := lhClient.LonghornV1beta1().Volumes(namespace).Update(context.TODO(), v, metav1.UpdateOptions{})
```

File: upgrade/v102to110/upgrade.go [hunk 5]
```diff
@@ -145,7 +145,7 @@ func upgradeReplicas(namespace string, lhClient *lhclientset.Clientset) (err err
 		return errors.Wrapf(err, "failed to list all existing Longhorn replicas during the replica upgrade")
 	}
 
-	for _, r := range replicaList.Items {
+	for i, r := range replicaList.Items {
```
Comment (shuo-wu) [inline upgrade/v102to110/upgrade.go:148]: Similar here:
```suggestion
for i := range replicaList.Items {
	r := &replicaList.Items[i]
```

File: upgrade/v110to111/upgrade.go [hunk 6]
```diff
@@ -131,7 +131,7 @@ func upgradeLonghornNodes(namespace string, lhClient *lhclientset.Clientset) (er
 		return errors.Wrapf(err, upgradeLogPrefix+"failed to list all existing Longhorn nodes")
 	}
 
-	for _, node := range nodeList.Items {
+	for i, node := range nodeList.Items {
```
Comment (shuo-wu) [inline upgrade/v110to111/upgrade.go:134]: Similar here:
```suggestion
for i := range nodeList.Items {
	node := &nodeList.Items[i]
```

File: upgrade/v110to120/upgrade.go [hunk 7]
```diff
@@ -52,22 +52,22 @@ func upgradeInstanceManagers(namespace string, lhClient *lhclientset.Clientset,
 	if err != nil {
 		return err
 	}
-	for _, imPod := range imPodList {
+	for i, imPod := range imPodList {
```
Comment (shuo-wu) [inline upgrade/v110to120/upgrade.go:55]: Similar here:
```suggestion
for i := range imPodList.Items {
	imPod := &imPodList.Items[i]
```

File: upgrade/v111to120/upgrade.go [hunk 8]
```diff
@@ -53,7 +53,7 @@ func upgradeBackingImages(namespace string, lhClient *lhclientset.Clientset) (er
 	if err != nil {
 		return err
 	}
-	for _, bi := range biList.Items {
+	for i, bi := range biList.Items {
```
Comment (shuo-wu) [inline upgrade/v111to120/upgrade.go:56]: Similar here:
```suggestion
for i := range biList.Items {
	bi := &biList.Items[i]
```

**Review by shuo-wu — CHANGES_REQUESTED** [review]

File: upgrade/v110to120/upgrade.go (hunk 7 above)
Comment (slayer321) [reply upgrade/v110to120/upgrade.go:55]: Here when I'm trying to add the `Items` field to `imPodList` .... I'm getting an error saying `Pod has no field or method Items`

**Comment by jenting** [comment]
@slayer321 are you willing to keep working on this PR?

**Comment by innobead** [comment]
Closing for long inactivity. @slayer321 feel free to continue this PR later on. Thank you.

**Pull request closed without merging**